File: frontend/lib/api/handoff.ts
export type HandoffDraftRequest = {
  pregnancy_status: string
  postpartum_hours: number | null
  danger_signs: string[]
  transport_mode: string
  blood_pressure: string
  heart_rate: number
  interventions_given: string[]
  clinician_note: string
}

export type HandoffDraftResponse = {
  draft: string
}

const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL

export async function generateHandoffDraft(
  payload: HandoffDraftRequest
): Promise<HandoffDraftResponse> {
  if (!API_BASE_URL) {
    throw new Error("NEXT_PUBLIC_API_BASE_URL is not set.")
  }

  const res = await fetch(`${API_BASE_URL}/api/handoff/draft`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  })

  const data = await res.json()

  if (!res.ok) {
    throw new Error(data?.detail || "Failed to generate handoff draft.")
  }

  return data
}

File: backend/app/api/routes/handoff.py
import logging

from fastapi import APIRouter, HTTPException

from app.schemas.handoff import HandoffDraftRequest
from app.services.watsonx_service import (
    build_local_handoff_draft,
    get_watsonx_service,
)

router = APIRouter(prefix="/api/handoff", tags=["handoff"])
logger = logging.getLogger(__name__)

GENERIC_HANDOFF_ERROR = "Could not generate handoff draft."
EXPECTED_HANDOFF_FALLBACK_ERRORS: tuple[type[Exception], ...] = (
    RuntimeError,
    OSError,
    ValueError,
    TypeError,
    KeyError,
    AttributeError,
    ImportError,
)


def _handoff_kwargs(payload: HandoffDraftRequest):
    return {
        "pregnancy_status": payload.pregnancy_status,
        "postpartum_hours": payload.postpartum_hours,
        "danger_signs": payload.danger_signs,
        "transport_mode": payload.transport_mode,
        "blood_pressure": payload.blood_pressure,
        "heart_rate": payload.heart_rate,
        "interventions_given": payload.interventions_given,
        "clinician_note": payload.clinician_note,
    }


def _generate_reviewable_draft(payload: HandoffDraftRequest) -> str:
    try:
        service = get_watsonx_service()
        return service.generate_handoff_draft(**_handoff_kwargs(payload))
    except EXPECTED_HANDOFF_FALLBACK_ERRORS as exc:
        logger.exception(
            "watsonx handoff generation failed on /api/handoff/draft (%s); using local fallback",
            exc.__class__.__name__,
        )
        return build_local_handoff_draft(**_handoff_kwargs(payload))


@router.post("/draft")
def generate_handoff_draft(payload: HandoffDraftRequest):
    try:
        return {"draft": _generate_reviewable_draft(payload)}
    except Exception as exc:
        logger.exception("Handoff draft generation failed")
        raise HTTPException(status_code=500, detail=GENERIC_HANDOFF_ERROR) from exc
